mono/sgen/sgen-marksweep.c

```diff
@@ -2339,7 +2357,31 @@ major_scan_card_table (CardTableScanType scan_type, ScanCopyContext ctx)
 		if (!has_references)
 			continue;
 
-		scan_card_table_for_block (block, scan_type, ctx);
+		if (scan_type == CARDTABLE_SCAN_GLOBAL) {
+			gpointer *card_start = (gpointer*) sgen_card_table_get_card_scan_address ((mword)MS_BLOCK_FOR_BLOCK_INFO (block));
+			int i;
+			for (i = 0; i < CARDS_PER_BLOCK / sizeof(gpointer); i++) {
+				if (card_start [i]) {
+					gboolean is_live = TRUE;
+					/*
+					 * After the start of the concurrent collections, blocks change state
+					 * to marking. We should not sweep it in that case. We can't race with
+					 * sweep start since we are in a nursery collection. Also avoid CAS-ing
+					 */
+					if (sweep_in_progress ()) {
+						is_live = ensure_block_is_checked_for_sweeping (__index, TRUE, NULL);
+					} else if (was_sweeping) {
+						/* Recheck in case sweep finished after dereferencing the slot */
+						is_live = *sgen_array_list_get_slot (&allocated_blocks, __index) != 0;
+					}
+					if (is_live)
+						scan_card_table_for_block (block, scan_type, ctx);
+					break;
+				}
+			}
+		} else {
+			scan_card_table_for_block (block, scan_type, ctx);
```

Comment:
In fact, there should be only one call to `scan_card_table_for_block()` in this outer loop, guarded by `is_live` that's `TRUE` by default and not changed in the `scan_type != CARDTABLE_SCAN_GLOBAL` case.